scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,385 @@
+        losses = []
+        for i, z in enumerate(latent_tensors):
+            cls_logits = nn.LogSoftmax(dim=1)(self.discriminator(z))
+
+            if predict_true_class:
+                cls_target = torch.zeros(
+                    n_classes, dtype=torch.float32, device=z.device
+                )
+                cls_target[i] = 1.0
+            else:
+                cls_target = torch.ones(
+                    n_classes, dtype=torch.float32, device=z.device
+                ) / (n_classes - 1)
+                cls_target[i] = 0.0
+
+            l_soft = cls_logits * cls_target
+            cls_loss = -l_soft.sum(dim=1).mean()
+            losses.append(cls_loss)
+
+        if return_details:
+            return losses
+
+        total_loss = torch.stack(losses).sum()
+        return total_loss
+
+    def loss(
+        self, tensors: Iterable[torch.Tensor], return_details: bool = False
+    ) -> Union[torch.Tensor, Tuple[List[torch.Tensor], List[torch.Tensor]]]:
+        """
+        Compute the loss of vae (reconstruction + kl_divergence)
+
+        :param tensors: Tensors of observations for each dataset
+        :param return_details: Boolean used to inspect the loss values, return detailed loss for each dataset
+        :return: scalar loss if return_details is False, else tuple (reconstruction_loss, kl_loss)
+        """
+        reconstruction_losses = []
+        kl_divergences = []
+        losses = []
+        total_batch_size = 0
+        for (i, (sample_batch, l_mean, l_var, batch_index, labels, *_)) in enumerate(
+            tensors
+        ):
+            reconstruction_loss, kl_divergence = self.model(
+                sample_batch, l_mean, l_var, batch_index, mode=i
+            )
+            loss = torch.mean(
+                reconstruction_loss + self.kl_weight * kl_divergence
+            ) * sample_batch.size(0)
+            total_batch_size += sample_batch.size(0)
+            losses.append(loss)
+            if return_details:
+                reconstruction_losses.append(reconstruction_loss.mean())
+                kl_divergences.append(kl_divergence.mean())
+
+        if return_details:
+            return reconstruction_losses, kl_divergences
+
+        averaged_loss = torch.stack(losses).sum() / total_batch_size
+        return averaged_loss
+
+    def get_discriminator_confusion(self) -> np.ndarray:
+        confusion = []
+        for i, posterior in enumerate(self.all_dataset):
+            data = torch.from_numpy(posterior.gene_dataset.X)
+            if self.use_cuda:
+                data.cuda()
+
+            z = self.model.sample_from_posterior_z(data, mode=i, deterministic=True)
+            cls_z = nn.Softmax(dim=1)(self.discriminator(z)).detach()
+
+            if self.use_cuda:
+                cls_z = cls_z.cpu()
+            cls_z = cls_z.numpy()
+
+            row = cls_z.mean(axis=0)
+            confusion.append(row)
+        return np.array(confusion)
+
+    def get_loss_magnitude(
```

Comment:
What is the meaning of magnitude here? Why not `get_total_loss` or `get_aggregate_loss` ?

Reply:
I introduced this function to track the different losses of this model: 
- reconstruction
- kl divergences
- adversarial
The main goal is to make sure they have the same order of magnitude (so that you don't have a losses that are actually never optimized because too small), so I called the function that way.
Does it make sense, or do you have another name in mind?

Reply:
So you don't use this for optimization purposes?
Maybe add a docstring then?

Reply:
Nope, all the `get_...` functions are for user inspection only.
I added docstrings